File: primes.py
from bisect import bisect_left
import sys

class Primes(object):
    def __init__(self):
        self._primes = [2]
        self._max = 2

    def is_prime(self, n):
        while self._max * self._max < n:
            self.expand()

        if self._max >= n:
            return self._primes[bisect_left(self._primes, n)] == n
        
        for i in self._primes:
            if i * i > n:
                return True
            if n % i == 0:
                return False
        return True

    def expand(self):
        if self.is_prime(self._max + 1):
            self._primes.append(self._max + 1)
        self._max += 1

primes = Primes()
for i in xrange(2, int(sys.argv[1])):
    if primes.is_prime(i):
        print i

File: primes.cc
#include <algorithm>
#include <iostream>
#include <stdint.h>
#include <vector>

class Primes {
  public:
    Primes();
    bool isPrime(uint32_t n) const;

  private:
    bool isPrimeInternal(uint32_t n);
    void expand();

    std::vector<uint32_t> primes;
    uint32_t max;
};

Primes::Primes() {
  primes.push_back(2);
  max = 2;
}

void Primes::expand() {
  if (isPrime(max + 1)) {
    primes.push_back(max + 1);
  }
  max++;
}

bool Primes::isPrime(uint32_t n) const {
  return const_cast<Primes *>(this)->isPrimeInternal(n);
}

bool Primes::isPrimeInternal(uint32_t n) {
  while (max * max < n) {
    expand();
  }

  if (max >= n) {
    return std::binary_search(primes.begin(), primes.end(), n);
  } else {
    for (std::vector<uint32_t>::iterator it = primes.begin(); it != primes.end(); it++) {
      uint32_t i = *it;
      if (i * i > n) {
        return true;
      } else if (n % i == 0) {
        return false;
      }
    }
    return true;
  }
}

int main(int argc, char** argv) {
  uint32_t n = atoi(argv[1]);
  Primes primes;
  for (uint32_t i = 1; i <= n; i++) {
    if (primes.isPrime(i)) {
      std::cout << i << std::endl;
    }
  }
}
